### ISSUE TITLE
Jupyter kernel on vscode gives error about variables declared on above cells

### ISSUE THREAD

**Issue by ceifa** [denoland/vscode_deno#932]
**To Reproduce**

1. Open deno kernel on vscode
2. Create a variable in the first cell
3. Use it on the second one

**Expected behavior**

None errors

**Screenshots**

<img width="1331" alt="image" src="https://github.com/denoland/deno/assets/26205666/c03637d0-8384-4d62-a295-b21c443d014e">

**Versions**

vscode: 1.82.2 deno: 1.37 extension: v3.23.1

@bartlomieju

**Comment by bartlomieju** [comment]
@nayeemrmn could you take a look what's going on here?

**Comment by nayeemrmn** [comment]
@ceifa I don't get intellisense at all. Is there some setting or supplementary extension you have installed that I can use to reproduce this?

**Comment by amunger** [comment]
This will happen in the deno language server isn't being used - try adding `"deno.enable": true` to see if the error goes away.

It looks like you can also put your notebooks in a specific directory and add that directory to `deno.enablePaths` as recommended in the setting description.

**Comment by amunger** [comment]
But I would say it should be safe for the extension to always use the deno language server for a notebook that has a deno kernel selected.

**Comment by ceifa** [comment]
> But I would say it should be safe for the extension to always use the deno language server for a notebook that has a deno kernel selected.

I agree. Specifically because I don't want deno to be enabled on the global settings and I also don't want to save the notebook file sometimes.

**Comment by ghing** [comment]
I'm also seeing this issue and have `"deno.enable": true` in my `.vscode/settings.json`.

**Comment by scarf005** [comment]
i'm wondering how this could be solved implementation-wise. my naive first impression was:

1. join all typescript cells
2. send the joined cell into tsc
3. parse tsc diagnostics and re-map line numbers into splitted cells

**Comment by nayeemrmn** [comment]
I forgot to give an update here, this is blocked by https://github.com/gluon-lang/lsp-types/pull/268 which implements this API in the protocol https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notebookDocument_synchronization. This provides the minimum of getting the accurate order of the cells etc.

After that it's still going to be difficult and we'd likely have to do what @scarf005 proposed above. On the surface this isn't a problem with VSCode's built-in node TS server because it has notebook cells in script mode, where all vars are global. We might alternatively end up doing something similar.

**Comment by bartlomieju** [comment]
Would this PR help somehow? https://github.com/denoland/deno/pull/21518

**Comment by pekam** [comment]
> I'm also seeing this issue and have `"deno.enable": true` in my `.vscode/settings.json`.

Same here. The Deno APIs are enabled as shown by this screenshot:
![image](https://github.com/denoland/vscode_deno/assets/28533233/eaaae4d4-f25e-4dbe-868f-775f77ef69a6)

Jupyter support is the only reason why I started using Deno, and this is issue is a blocker for me.

**Comment by Hexodus** [comment]
What a pitty, I was so excited to try out Deno inside of jupyter notebooks but having variables that don't remember their values makes it a no go. And yes, Deno is enabled in this workspace.
![image](https://github.com/denoland/vscode_deno/assets/243039/5500899c-52cc-49cc-88dc-ab10325aadac)

**Comment by sigmaSd** [comment]
Just as a reminder, the code does work as expected, this issue is just about the lsp giving wrong diagnostics, but if you run the cell it will work.

**Comment by vanb** [comment]
> Just as a reminder, the code does work as expected, this issue is just about the lsp giving wrong diagnostics, but if you run the cell it will work.

Unfortunately, even if it does execute, it's frustrating to lose all type information and be downgraded to `any` after the first cell. So, no code completion or type checking, effectively losing all benefits of typescript. Any ETA on when this might be fixed?

**Comment by matijagrcic** [comment]
This would be great to have. 

I've read the comments: 

https://github.com/denoland/vscode_deno/issues/932#issuecomment-1848795484

> I forgot to give an update here, this is blocked by [gluon-lang/lsp-types#268](https://github.com/gluon-lang/lsp-types/pull/268) which implements this API in the protocol https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notebookDocument_synchronization. This provides the minimum of getting the accurate order of the cells etc.
> 
> After that it's still going to be difficult and we'd likely have to do what @scarf005 proposed above. On the surface this isn't a problem with VSCode's built-in node TS server because it has notebook cells in script mode, where all vars are global. We might alternatively end up doing something similar.



https://github.com/denoland/deno/pull/21518#issuecomment-2094338519

> I am no longer working on this PR.
> 
> If I remember correctly, the Deno language server does not have proper notebook support but treats every cell like a separate document. This is the default fallback mechanism for notebooks if the language server does not support notebooks. That behavior causes some problems; for example, the type of variables from the previous cell is unknown in the subsequent cells. Fixing this would require the language server to treat all cells as a single file, but that would come with other issues, as the cells are not necessarily executed from top to bottom. However, that would still be more useful than the current implementation. Afaik, no language server for a statically typed language currently has _good_ support for notebooks.
> 
> The language server protocol supports notebooks, but we need `tower-lsp` to do the same. I opened a PR over there ([ebkalderon/tower-lsp#401](https://github.com/ebkalderon/tower-lsp/pull/401)), but that is blocked by [gluon-lang/lsp-types#266](https://github.com/gluon-lang/lsp-types/pull/266), which is blocked by [gluon-lang/lsp-types#266](https://github.com/gluon-lang/lsp-types/pull/266).
> 
> I am closing this PR because I have given up on using deno for notebooks.

So wondering can [import_maps](https://docs.deno.com/runtime/manual/basics/import_maps) be used to propagate type information across the cells?

So far i'm leveraging [deps.ts](https://docs.deno.com/runtime/tutorials/manage_dependencies) and it's not ideal as it still complains about unused values etc, but at least i'm getting the intelliSense.

>The standard practice for solving this problem in Deno is to create a deps.ts file. All required remote dependencies are referenced in this file and the required methods and classes are re-exported. The dependent local modules then reference the deps.ts rather than the remote dependencies.

**Comment by planetoryd** [comment]
@nayeemrmn gluon PR has been merged can this be resumed

**Comment by planetoryd** [comment]
on top of that it'd be nice to have chrome-devtools-like object inspector (that expands as you click on the items) in notebook cell output

**Comment by matijagrcic** [comment]
> on top of that it'd be nice to have chrome-devtools-like object inspector (that expands as you click on the items) in notebook cell output

You can already use rich display for that if you want https://deno.land/x/display@v1.1.2 and build custom formatters.

**Comment by jarrodcolburn** [comment]
The irony is that because the `jupyter` extension works fine for `typescript` by itself and won't warn of variables, and gives types hints.,
When the `deno extension` is `disabled` (`jupyter` is `enabled`) it works fine.

So my temporary workaround is... 
* Create directory in project root to store `.ipbyn` files (example: `./notebook`)
* disable deno extension for that directory in workspace's `settings.json` example:
```json
{
    "deno.enable":true, 
    "deno.disablePaths": [
        "./notebooks"
    ]
}
```
* Only rely on `jupyter extension` 
![image](https://github.com/user-attachments/assets/3bc554d5-744b-442f-80b0-61d36bd27101)

**Comment by redking00** [comment]
Hello!,
Check out [my fork](https://github.com/redking00/vscode-nbts). It adds support for Deno LSP on notebooks. It also defines a new notebook type (*.nb.ts). This way you can use "import" between notebooks.
Lastly, it also adds a notebook driver for vscode, so you can use deno's jupyter kernel in vscode without Python.
You just have to install the vsix file of the extension, look at the releases. No need for Jupyter or any other extension (deno must be in the path, or configuration setted).
Remember deno.enable:true in the workspace that contains the notebooks. The best way is to use the command "Deno: Initialize workspace configuration" (adds some more configuration besides deno.enable). 
Rename any .ts file to .nb.ts to get started. You can also start with an empty file: "touch test.nb.ts".
Leave me a comment if you like it.
Greetings.

**Comment by planetoryd** [comment]
![image](https://github.com/user-attachments/assets/849796c7-9224-464b-8bbc-1bdc1cd876d8)

@redking00 

thing is awesome but it crashes a lot and here in the picture it is missing an item

**Comment by planetoryd** [comment]
![image](https://github.com/user-attachments/assets/67097338-3743-4bbc-b75d-f96aab31f996)

@redking00 pretty good but autocompletion doesnt work for object `x`

**Comment by planetoryd** [comment]
this might be of some value  

```
error: Uncaught (in promise) TypeError: Cannot set properties of null (setting 'data')
          data.result.data = [];
                          ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:288:27)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Info  - 5:58:54 AM] Connection to server got closed. Server will restart.
Starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
error: Uncaught (in promise) TypeError: Cannot set properties of null (setting 'data')
          data.result.data = [];
                          ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:288:27)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Error - 5:58:55 AM] The Deno Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

**Comment by scarf005** [comment]
@redking00 that's awesome, do you plan to mainline the changes from your fork soon?

**Comment by planetoryd** [comment]
![image](https://github.com/user-attachments/assets/7d7237c7-271e-4635-b10a-32d679f35c2f)
![image](https://github.com/user-attachments/assets/35f51969-5978-4728-882f-186c1639e3d9)

<details>

<summary>"Deno Language Server" logs </summary>

```
starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
successfully cancelled request with ID: 26
successfully cancelled request with ID: 27
successfully cancelled request with ID: 28
error: Uncaught (in promise) TypeError: Cannot read properties of null (reading 'range')
          data.result.range.start.line -= startLine;
                      ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:250:23)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Info  - 5:58:37 AM] Connection to server got closed. Server will restart.
Starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
error: Uncaught (in promise) TypeError: Cannot read properties of null (reading 'range')
          data.result.range.start.line -= startLine;
                      ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:250:23)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Info  - 5:58:39 AM] Connection to server got closed. Server will restart.
Error retrieving config tasks: Error: Pending response rejected since connection got disposed
Starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
error: Uncaught (in promise) TypeError: Cannot set properties of null (setting 'data')
          data.result.data = [];
                          ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:288:27)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Info  - 5:58:53 AM] Connection to server got closed. Server will restart.
Starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
successfully cancelled request with ID: 15
error: Uncaught (in promise) TypeError: Cannot set properties of null (setting 'data')
          data.result.data = [];
                          ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:288:27)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Info  - 5:58:54 AM] Connection to server got closed. Server will restart.
Starting Deno language server...
  version: 2.0.4 (release, x86_64-unknown-linux-gnu)
  executable: /home/phy/.cargo/bin/deno
Connected to "Visual Studio Code - Insiders" 1.96.0-insider
Enabling import suggestions for: https://deno.land
Refreshing configuration tree...
  Resolved Deno configuration file: "file:///home/phy/drugbankdb/deno.json"
  Resolved lockfile: "file:///home/phy/drugbankdb/deno.lock"
Server ready.
error: Uncaught (in promise) TypeError: Cannot set properties of null (setting 'data')
          data.result.data = [];
                          ^
    at onDenoResponse (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:288:27)
    at StreamMessageReader.callback (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/lsp-server/notebook-lsp.ts:358:5)
    at Object.thunk (file:///home/phy/.vscode-insiders/extensions/redking00.vscode-nbts-0.0.2/client/node_modules/vscode-jsonrpc/lib/common/messageReader.js:205:26)
    at eventLoopTick (ext:core/01_core.js:214:9)
[Error - 5:58:55 AM] The Deno Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

</details>

**Comment by redking00** [comment]
@ple1n Thanks for the report. Just fixed, can you test version 0.0.3?

**Comment by redking00** [comment]
> @redking00 that's awesome, do you plan to mainline the changes from your fork soon?

thanks @scarf005, just starting for now. Not sure it's useful to anyone yet.

**Comment by ple1n** [comment]
![image](https://github.com/user-attachments/assets/4c12b7f4-89bd-445b-b74b-895767c63865)

**Comment by ple1n** [comment]
reinstalled extension by manually doing uninstall and reloaded window, retried, problem still exists 

![image](https://github.com/user-attachments/assets/d4cdba06-0883-4f85-845d-e422bbc50621)


also I have official Deno extension installed. it doesnt conflict right

**Comment by ple1n** [comment]
disabled it. problem persists 

![image](https://github.com/user-attachments/assets/c7f44d49-677d-4277-a8e7-e8e8b95fd42f)

![image](https://github.com/user-attachments/assets/99129676-9851-40a3-afa4-bce329f9e53b)

on 0.0.4 @redking00

**Comment by redking00** [comment]
@ple1n Just tested on a fresh 1.96.0-insiders (on linux manjaro with deno 2.0.4 on path) and no problem. Is deno on your path? what configuration (user/workspace) do you have? Can you share a minimal example?

**Comment by ple1n** [comment]
![image](https://github.com/user-attachments/assets/1a288612-370f-4e46-87ec-8d334b47c5c8)

(ms-build)

yes on my path. 

```
➜  drugbankdb git:(master) ✗ deno -V 
deno 2.0.4
➜  drugbankdb git:(master) ✗ which deno 
/usr/local/bin/deno
➜  drugbankdb git:(master) ✗ ls -lh `which deno`
lrwxrwxrwx. 1 root root 25 Nov  4 05:19 /usr/local/bin/deno -> /home/phy/.cargo/bin/deno
```


```json
{
    "telemetry.telemetryLevel": "off",
    "jupyter.askForKernelRestart": false,
    "git.openRepositoryInParentFolders": "never",
    "notebook.scrolling.revealNextCellOnExecute": "none",
    "extensions.ignoreRecommendations": true,
    "jupyter.widgetScriptSources": [
        "jsdelivr.com",
        "unpkg.com"
    ],
    "[typescript]": {
        "editor.defaultFormatter": "vscode.typescript-language-features"
    },
    "notebook.lineNumbers": "on",
    "deno.path": "",
    "deno.internalDebug": true,
    "deno.trace.server": "messages"
}
```

user config, workspace config empty

**Comment by redking00** [comment]
@ple1n Can you check with "deno.path":"/home/phy/.cargo/bin/deno" ? Thanks

**Comment by ple1n** [comment]
why does deno exit. it crashed? and what is that object serialize it and print it ig.

**Comment by ple1n** [comment]
![image](https://github.com/user-attachments/assets/c3330d25-1681-450d-aba2-32f7d19e9918)

deno is compiled from source

**Comment by redking00** [comment]
@ple1n 
Seems deno is exiting. The object is the spawned process. My guess is it's a problem opening the zeromq sockets.  Can you try this?

Save this as _connection_file_
`
{
 "ip": "127.0.0.1",
 "iopub_port": 40885,
 "shell_port": 40886,
 "control_port": 40887,
 "hb_port": 40888,
 "stdin_port": 40889,
 "transport": "tcp",
 "signature_scheme": "sha256",
 "key": "ba497f9f-1606-4dbd-aa45-cfc8a70ffc57"
}
`

Then launch 
`deno jupyter --kernel --conn connection_file`

Is deno exiting after that?

**Comment by ple1n** [comment]
I cloned your repo and set it up. seems source maps are not working cant debug it. 

![image](https://github.com/user-attachments/assets/580075b4-5b97-4cf1-a7d9-6de8f52a7031)

btw is this gonna do anything js console is better for objects

**Comment by ple1n** [comment]
![image](https://github.com/user-attachments/assets/58825199-ab6b-4d3d-99ea-73cdc8a9dc6e)

@redking00

**Comment by ple1n** [comment]
```
➜  ~ deno jupyter --kernel --conn connfile 
Warning "deno jupyter" is unstable and might change in the future.
error: Failed to acquire startup data
```

**Comment by ple1n** [comment]
I killed that process. It seems to be some left-over process that occupied the ports. And it worked

**Comment by redking00** [comment]
I'm glad to hear that! @ple1n  Thanks for your interest

**Comment by ple1n** [comment]
I mean how should the auto completion for that object be done. seems like a non trivial issue. infer the typing based on the object and send it back to lsp or such? and is that within the scope of your fork/pr / requiring deno-team intervention

**Comment by redking00** [comment]
> I mean how should the auto completion for that object be done. seems like a non trivial issue. infer the typing based on the object and send it back to lsp or such? and is that within the scope of your fork/pr / requiring deno-team intervention

I have no idea how something like that could be done. It certainly wasn't on my to-do list for this project, but I'm open to helping in any way I can.

**Comment by ple1n** [comment]
It conflicts with other instances of the extension, ie I cant run multiple notebooks

**Comment by redking00** [comment]
@ple1n Just updated to v0.0.5. Can you check it?. I tested two VSCode instances with no problem.

**Comment by ple1n** [comment]
> @ple1n There is no problem running multiple notebooks on the same VSCode instance so I assume you mean multiple VSCode instances. It's probably the same problem with the ports. I have to add a random initial offset.

Yes exactly 

![image](https://github.com/user-attachments/assets/d011528b-6d15-4c46-9847-c61c98c051bb)

it pops up the error every second I write the sql statement (obviosuly it was kinda sending the sql to LSP for some reason). I cant suppress the notification it gets annoying

**Comment by ple1n** [comment]
no. it errors elsewhere too

**Comment by ple1n** [comment]
```

[Trace - 5:10:19 AM] Received response 'textDocument/codeAction - (5017)' in 8ms. Request failed: Invalid request (-32600).
[Error - 5:10:19 AM] Request textDocument/codeAction failed.
  Message: Invalid request
  Code: -32600 
[Trace - 5:10:19 AM] Sending request 'textDocument/codeAction - (5018)'.
Failed to request to tsserver tsc error: TypeError: Cannot read properties of undefined (reading 'kind')
    at isVariableDeclaration (ext:deno_tsc/00_typescript.js:30518:15)
    at isSingleVariableDeclaration (ext:deno_tsc/00_typescript.js:146058:10)
    at tryGetFunctionFromVariableDeclaration (ext:deno_tsc/00_typescript.js:146061:8)
    at getFunctionInfo (ext:deno_tsc/00_typescript.js:146046:16)
    at Object.getRefactorActionsToConvertFunctionExpressions [as getAvailableActions] (ext:deno_tsc/00_typescript.js:145971:16)
    at ext:deno_tsc/00_typescript.js:143671:222
    at flatMapIterator (ext:deno_tsc/00_typescript.js:2545:19)
    at flatMapIterator.next (<anonymous>)
    at arrayFrom (ext:deno_tsc/00_typescript.js:3038:14)
    at Object.getApplicableRefactors (ext:deno_tsc/00_typescript.js:143669:10)
For request: [3193, "getApplicableRefactors", ["file:///home/phy/drugbankdb/peek1.nb.ts",{"pos":953,"end":953},{"quotePreference":"double","includeCompletionsForModuleExports":true,"includeCompletionsForImportStatements":true,"includeCompletionsWithSnippetText":true,"includeAutomaticOptionalChainCompletions":true,"includeCompletionsWithInsertText":true,"includeCompletionsWithClassMemberSnippets":true,"includeCompletionsWithObjectLiteralMethodSnippets":true,"useLabelDetailsInCompletionEntries":true,"allowIncompleteCompletions":true,"importModuleSpecifierPreference":"shortest","importModuleSpecifierEnding":"index","allowTextChangesInNewFiles":true,"providePrefixAndSuffixTextForRename":true,"provideRefactorNotApplicableReason":true,"jsxAttributeCompletionStyle":"auto","includeInlayParameterNameHints":"none","includeInlayParameterNameHintsWhenArgumentMatchesName":false,"includeInlayFunctionParameterTypeHints":false,"includeInlayVariableTypeHints":false,"includeInlayVariableTypeHintsWhenTypeMatchesName":false,"includeInlayPropertyDeclarationTypeHints":false,"includeInlayFunctionLikeReturnTypeHints":false,"includeInlayEnumMemberValueHints":false,"autoImportFileExcludePatterns":[],"interactiveInlayHints":true,"preferTypeOnlyAutoImports":false},"implicit",""], "file:///home/phy/drugbankdb/", null]
[Trace - 5:10:19 AM] Received response 'textDocument/codeAction - (5018)' in 6ms. Request failed: Invalid request (-32600).
[Error - 5:10:19 AM] Request textDocument/codeAction failed.
  Message: Invalid request
  Code: -32600 
[Trace - 5:10:20 AM] Sending request 'textDocument/inlayHint - (5019)'.
[Trace - 5:10:20 AM] Received response 'textDocument/inlayHint - (5019)' in 13ms.
[Trace - 5:10:40 AM] Sending request 'textDocument/codeLens - (5020)'.
[Trace - 5:10:40 AM] Received response 'textDocument/codeLens - (5020)' in 6ms.
[Trace - 5:10:42 AM] Sending request 'textDocument/codeAction - (5021)'.
Failed to request to tsserver tsc error: TypeError: Cannot read properties of undefined (reading 'kind')
    at isVariableDeclaration (ext:deno_tsc/00_typescript.js:30518:15)
    at isSingleVariableDeclaration (ext:deno_tsc/00_typescript.js:146058:10)
    at tryGetFunctionFromVariableDeclaration (ext:deno_tsc/00_typescript.js:146061:8)
    at getFunctionInfo (ext:deno_tsc/00_typescript.js:146046:16)
    at Object.getRefactorActionsToConvertFunctionExpressions [as getAvailableActions] (ext:deno_tsc/00_typescript.js:145971:16)
    at ext:deno_tsc/00_typescript.js:143671:222
    at flatMapIterator (ext:deno_tsc/00_typescript.js:2545:19)
    at flatMapIterator.next (<anonymous>)
    at arrayFrom (ext:deno_tsc/00_typescript.js:3038:14)
    at Object.getApplicableRefactors (ext:deno_tsc/00_typescript.js:143669:10)
For request: [3194, "getApplicableRefactors", ["file:///home/phy/drugbankdb/peek1.nb.ts",{"pos":1027,"end":1027},{"quotePreference":"double","includeCompletionsForModuleExports":true,"includeCompletionsForImportStatements":true,"includeCompletionsWithSnippetText":true,"includeAutomaticOptionalChainCompletions":true,"includeCompletionsWithInsertText":true,"includeCompletionsWithClassMemberSnippets":true,"includeCompletionsWithObjectLiteralMethodSnippets":true,"useLabelDetailsInCompletionEntries":true,"allowIncompleteCompletions":true,"importModuleSpecifierPreference":"shortest","importModuleSpecifierEnding":"index","allowTextChangesInNewFiles":true,"providePrefixAndSuffixTextForRename":true,"provideRefactorNotApplicableReason":true,"jsxAttributeCompletionStyle":"auto","includeInlayParameterNameHints":"none","includeInlayParameterNameHintsWhenArgumentMatchesName":false,"includeInlayFunctionParameterTypeHints":false,"includeInlayVariableTypeHints":false,"includeInlayVariableTypeHintsWhenTypeMatchesName":false,"includeInlayPropertyDeclarationTypeHints":false,"includeInlayFunctionLikeReturnTypeHints":false,"includeInlayEnumMemberValueHints":false,"autoImportFileExcludePatterns":[],"interactiveInlayHints":true,"preferTypeOnlyAutoImports":false},"implicit",""], "file:///home/phy/drugbankdb/", null]
[Trace - 5:10:42 AM] Received response 'textDocument/codeAction - (5021)' in 8ms. Request failed: Invalid request (-32600).
[Error - 5:10:42 AM] Request textDocument/codeAction failed.
  Message: Invalid request
  Code: -32600 
[Trace - 5:10:42 AM] Sending request 'textDocument/inlayHint - (5022)'.
[Trace - 5:10:42 AM] Received response 'textDocument/inlayHint - (5022)' in 5ms.
[Trace - 5:10:42 AM] Sending request 'textDocument/foldingRange - (5023)'.
[Trace - 5:10:42 AM] Received response 'textDocument/foldingRange - (5023)' in 6ms.
[Trace - 5:10:42 AM] Sending request 'textDocument/inlayHint - (5024)'.
[Trace - 5:10:42 AM] Sending request 'textDocument/codeAction - (5025)'.
[Trace - 5:10:42 AM] Received response 'textDocument/inlayHint - (5024)' in 5ms.
Failed to request to tsserver tsc error: TypeError: Cannot read properties of undefined (reading 'kind')
    at isVariableDeclaration (ext:deno_tsc/00_typescript.js:30518:15)
    at isSingleVariableDeclaration (ext:deno_tsc/00_typescript.js:146058:10)
    at tryGetFunctionFromVariableDeclaration (ext:deno_tsc/00_typescript.js:146061:8)
    at getFunctionInfo (ext:deno_tsc/00_typescript.js:146046:16)
    at Object.getRefactorActionsToConvertFunctionExpressions [as getAvailableActions] (ext:deno_tsc/00_typescript.js:145971:16)
    at ext:deno_tsc/00_typescript.js:143671:222
    at flatMapIterator (ext:deno_tsc/00_typescript.js:2545:19)
    at flatMapIterator.next (<anonymous>)
    at arrayFrom (ext:deno_tsc/00_typescript.js:3038:14)
    at Object.getApplicableRefactors (ext:deno_tsc/00_typescript.js:143669:10)
For request: [3196, "getApplicableRefactors", ["file:///home/phy/drugbankdb/peek1.nb.ts",{"pos":1035,"end":1035},{"quotePreference":"double","includeCompletionsForModuleExports":true,"includeCompletionsForImportStatements":true,"includeCompletionsWithSnippetText":true,"includeAutomaticOptionalChainCompletions":true,"includeCompletionsWithInsertText":true,"includeCompletionsWithClassMemberSnippets":true,"includeCompletionsWithObjectLiteralMethodSnippets":true,"useLabelDetailsInCompletionEntries":true,"allowIncompleteCompletions":true,"importModuleSpecifierPreference":"shortest","importModuleSpecifierEnding":"index","allowTextChangesInNewFiles":true,"providePrefixAndSuffixTextForRename":true,"provideRefactorNotApplicableReason":true,"jsxAttributeCompletionStyle":"auto","includeInlayParameterNameHints":"none","includeInlayParameterNameHintsWhenArgumentMatchesName":false,"includeInlayFunctionParameterTypeHints":false,"includeInlayVariableTypeHints":false,"includeInlayVariableTypeHintsWhenTypeMatchesName":false,"includeInlayPropertyDeclarationTypeHints":false,"includeInlayFunctionLikeReturnTypeHints":false,"includeInlayEnumMemberValueHints":false,"autoImportFileExcludePatterns":[],"interactiveInlayHints":true,"preferTypeOnlyAutoImports":false},"implicit",""], "file:///home/phy/drugbankdb/", null]
[Trace - 5:10:42 AM] Sending request 'textDocument/codeLens - (5026)'.
[Trace - 5:10:42 AM] Received response 'textDocument/codeAction - (5025)' in 7ms. Request failed: Invalid request (-32600).
[Error - 5:10:42 AM] Request textDocument/codeAction failed.
  Message: Invalid request
  Code: -32600 
[Trace - 5:10:42 AM] Received response 'textDocument/codeLens - (5026)' in 4ms.
```

**Comment by ple1n** [comment]
and as predicted, by intuition. it disappears after reloading window

**Comment by redking00** [comment]
For those interested in a quick test on the lsp thing... [a codespaces template](https://github.com/redking00/vscode-nbts-codespaces).

@nayeemrmn , @bartlomieju, Any interest in this or do I keep the fork? Maybe it can be included as an opt-in setting for notebooks users until the real lsp is ready.

**Comment by ple1n** [comment]
check my project out https://github.com/planetoryd/drugbankdb 

I've been using this notebook feature with SurrealDB for data exploration. Its publicly available drugbank data you just need an account to download. 

I am not associated with anyone else its just for personal needs. 

The alternative is to use ELK stack, which does not work with graph data yet. Or use MongoDB stack, which did not work out for me, or has good features paywalled.

And I am NOT gonna use python+pylance for surrealdb. It will suck.

**Comment by ple1n** [comment]
![image](https://github.com/user-attachments/assets/2fd2acfb-b010-413c-8b78-149516ec3ffd)

**Comment by bartlomieju** [comment]
> For those interested in a quick test on the lsp thing... [a codespaces template](https://github.com/redking00/vscode-nbts-codespaces).
> 
> @nayeemrmn , @bartlomieju, Any interest in [this](https://github.com/redking00/vscode-nbts) or do I keep the fork? Maybe it can be included as an opt-in setting for notebooks users until the real lsp is ready.

Thanks @redking00, @nayeemrmn is probably 50% done with the LSP integration at this point, but we have a few more pressing issues to fix before then. I'm not super keen on bringing a lot of new functionality into the extension as we try to keep it slim and offload most of the work to the LSP so it's easily testable. So I'd say it might be better to keep the fork going for now.

**Comment by redking00** [comment]
@bartlomieju, @nayeemrmn, 
I love knowing that work is being done on the lsp. 
Being able to run the kernel without python seems to me to be something you should take into account for the future. 
And the serializer... I think it is very useful for maintaining dependencies between notebooks, especially in the field of literate programming.
In any case, I understand what you are saying. 
I will keep the fork.
I love your work! Keep it up!

**Comment by bartlomieju** [comment]
@redking00 I might be missing some crucial detail here - maybe you could open issues in the Deno repo with suggestions for the kernel and the serializer?

**Comment by redking00** [comment]
@bartlomieju Both are vscode stuff.

**Comment by phromo** [comment]
Since the discussion is very long, can anyone summarize the state of this issue? Is setting dino.enabled to true supposed to fix this problem, or is it expected to still exist using the denoland vscode extension?

**Comment by gnlow** [comment]
@phromo 
Current workaround is to use [DenoNBTS](https://github.com/redking00/vscode-nbts), which is a fork of vscode_deno.

**Comment by janckerchen** [comment]
What is the current progress of LSP?

**Comment by callmephilip** [comment]
> [@phromo](https://github.com/phromo) summarization: Current workaround is to use [DenoNBTS](https://github.com/redking00/vscode-nbts), which is a fork of vscode_deno.

yup. also, massive thanks to @redking00 - what a legend!

**Comment by matijagrcic** [comment]
Wow, amazing!
